File: packages/github_agents/src/github_agents/security/manager.py
"""Security manager for GitHub AI Agents."""

from datetime import datetime, timedelta
import json
import logging
import os
from pathlib import Path
import re
from typing import Any, Dict, List, Optional, Tuple

logger = logging.getLogger(__name__)


class SecurityManager:
    """Manages security for AI agent operations."""

    def __init__(self, agent_config: Optional[Any] = None, config_path: Optional[Path] = None) -> None:
        """Initialize security manager.

        Args:
            agent_config: AgentConfig instance to use for security settings
            config_path: Path to security config file (deprecated, use agent_config)
        """
        if agent_config:
            # Use security settings from AgentConfig
            security_config = agent_config.get_security_config()
            # Merge with defaults to ensure all required keys exist
            default_config = self._get_default_config()
            self.config = {**default_config, **security_config}
        else:
            # Fallback to loading from file
            self.config = self._load_config(config_path)
        self.rate_limit_tracker: Dict[str, List[datetime]] = {}

        # Initialize allowed users from config and environment
        self._init_allowed_users()

        # Initialize allowed repositories from environment
        self._init_allowed_repositories()

    def _get_default_config(self) -> dict:
        """Get default security configuration."""
        return {
            "enabled": True,
            "allow_list": ["AndrewAltimit", "github-actions[bot]", "dependabot[bot]"],
            "rate_limit_window_minutes": 60,
            "rate_limit_max_requests": 10,
            "allowed_repositories": [],
            "reject_message": "This AI agent only processes requests from authorized users.",
            "allowed_actions": [
                "issue_approved",
                "issue_close",
                "pr_approved",
                "issue_review",
                "pr_review",
                "issue_summarize",
                "pr_summarize",
                "issue_debug",
                "pr_debug",
            ],
            "triggers": {
                "approved": ["Approved"],
                "review": ["Review"],
                "close": ["Close"],
                "summarize": ["Summarize"],
                "debug": ["Debug"],
            },
        }

    def _load_config(self, config_path: Optional[Path]) -> dict:
        """Load security configuration."""
        default_config = self._get_default_config()

        if config_path and config_path.exists():
            try:
                with open(config_path, encoding="utf-8") as f:
                    loaded_config = json.load(f)
                    return {**default_config, **loaded_config.get("security", {})}
            except Exception as e:
                logger.warning("Failed to load security config: %s", e)

        return default_config

    def _init_allowed_users(self) -> None:
        """Initialize allowed users from config and environment variables."""
        # Start with users from config
        self._allowed_users: set[str] = set(self.config["allow_list"])

        # Add users from environment variable
        env_users = os.getenv("AI_AGENT_ALLOWED_USERS", "")
        if env_users:
            self._allowed_users.update(user.strip() for user in env_users.split(",") if user.strip())

        # Add repository owner if GITHUB_REPOSITORY is set
        github_repo = os.getenv("GITHUB_REPOSITORY", "")
        if github_repo and "/" in github_repo:
            owner = github_repo.split("/")[0]
            self._allowed_users.add(owner)

    def _init_allowed_repositories(self) -> None:
        """Initialize allowed repositories from environment variables."""
        # Start with repositories from config
        config_repos = self.config.get("allowed_repositories", [])

        # Check environment variable
        env_repos = os.getenv("AI_AGENT_ALLOWED_REPOS", "")
        if env_repos:
            # Environment variable takes precedence
            self.config["allowed_repositories"] = [repo.strip() for repo in env_repos.split(",") if repo.strip()]
        elif not config_repos:
            # If no repos specified, empty list means all allowed
            self.config["allowed_repositories"] = []

    @property
    def allowed_users(self) -> set[str]:
        """Get set of allowed users."""
        return self._allowed_users

    @property
    def allowed_actions(self) -> list[str]:
        """Get list of allowed actions."""
        return list(self.config["allowed_actions"])

    @property
    def triggers(self) -> dict[str, list[str]]:
        """Get trigger patterns."""
        return dict(self.config["triggers"])

    def parse_trigger_comment(self, text: str) -> Tuple[Optional[str], Optional[str]]:
        """Parse trigger comment and extract action and optional agent.

        Supports two formats:
        - [Action] - agent will be resolved from board or config
        - [Action][Agent] - explicit agent override

        Args:
            text: Comment text to parse

        Returns:
            Tuple of (action, agent) in lowercase. Agent may be None if not specified.
            Returns (None, None) if no valid trigger found.
        """
        if not text:
            return (None, None)

        # Pattern: [Action] with optional [Agent] - case insensitive
        pattern = r"\[([A-Za-z]+)\](?:\[([A-Za-z]+)\])?"
        match = re.search(pattern, text, re.IGNORECASE)

        if not match:
            return (None, None)

        action = match.group(1)
        agent = match.group(2)  # May be None if not specified

        # Validate action against allowed triggers (case-insensitive)
        valid_actions_lower = ["approved", "review", "close", "summarize", "debug"]
        if action.lower() not in valid_actions_lower:
            return (None, None)

        # Return lowercase (agent is None if not specified)
        return (action.lower(), agent.lower() if agent else None)

    def is_user_allowed(self, username: str) -> bool:
        """Check if user is authorized.

        Args:
            username: GitHub username

        Returns:
            True if authorized
        """
        if not self.config["enabled"]:
            return True

        return username in self._allowed_users

    def is_action_allowed(self, action: str) -> bool:
        """Check if action is allowed.

        Args:
            action: Action to check

        Returns:
            True if allowed
        """
        if not self.config["enabled"]:
            return True

        return action in self.config["allowed_actions"]

    def mask_secrets(self, text: str) -> str:
        """Mask secrets in text.

        Args:
            text: Text to mask secrets in

        Returns:
            Text with secrets masked
        """
        masked_text = text

        # Get list of environment variables to mask
        mask_vars = os.getenv("MASK_ENV_VARS", "")
        if not mask_vars:
            return masked_text

        # Mask each specified environment variable
        for var_name in mask_vars.split(","):
            var_name = var_name.strip()
            if not var_name:
                continue

            var_value = os.getenv(var_name, "")
            if var_value:
                masked_text = masked_text.replace(var_value, "***")

        return masked_text

    def get_trigger_regex(self) -> str:
        """Get regex pattern for valid triggers.

        Returns:
            Regex pattern string
        """
        # Pattern: [Action] with optional [Agent] where Action is one of the valid actions
        return r"\[(Approved|Review|Close|Summarize|Debug)\](?:\[([A-Za-z]+)\])?"

    def check_trigger_comment(self, issue_or_pr: Dict, _entity_type: str) -> Optional[Tuple[str, Optional[str], str]]:
        """Check for valid trigger in issue/PR comments.

        Args:
            issue_or_pr: Issue or PR data with comments
            _entity_type: "issue" or "pr"

        Returns:
            Tuple of (action, agent, username) if valid trigger found.
            Agent may be None if not specified in trigger (will be resolved from board).
            Returns None if no valid trigger found.
        """
        # Check issue/PR body first
        body = issue_or_pr.get("body", "")
        author = issue_or_pr.get("author", {}).get("login", "")

        action, agent = self.parse_trigger_comment(body)
        if action and self._is_user_authorized(author):
            return action, agent, author

        # Check comments
        for comment in issue_or_pr.get("comments", []):
            comment_body = comment.get("body", "")
            comment_author = comment.get("author", {}).get("login", "")

            action, agent = self.parse_trigger_comment(comment_body)
            if action and self._is_user_authorized(comment_author):
                return action, agent, comment_author

        return None

    def _is_user_authorized(self, username: str) -> bool:
        """Check if user is authorized.

        Args:
            username: GitHub username

        Returns:
            True if authorized
        """
        if not self.config["enabled"]:
            return True

        return username in self._allowed_users

    def check_rate_limit(self, username: str, action: str) -> bool:
        """Check if user has exceeded rate limit.

        Args:
            username: GitHub username
            action: Action being performed

        Returns:
            True if within rate limit
        """
        if not self.config["enabled"]:
            return True

        key = f"{username}:{action}"
        now = datetime.now()
        window = timedelta(minutes=self.config["rate_limit_window_minutes"])
        max_requests = self.config["rate_limit_max_requests"]

        # Clean old entries
        if key in self.rate_limit_tracker:
            self.rate_limit_tracker[key] = [t for t in self.rate_limit_tracker[key] if now - t < window]
        else:
            self.rate_limit_tracker[key] = []

        # Check limit
        if len(self.rate_limit_tracker[key]) >= max_requests:
            return False

        # Record request
        self.rate_limit_tracker[key].append(now)
        return True

    def is_repository_allowed(self, repository: str) -> bool:
        """Check if repository is allowed.

        Args:
            repository: Repository in format "owner/repo"

        Returns:
            True if allowed
        """
        if not self.config["enabled"]:
            return True

        allowed_repos = self.config["allowed_repositories"]
        if not allowed_repos:  # Empty list means all repos allowed
            return True

        return repository in allowed_repos

    # pylint: disable=unused-argument  # entity_type/entity_id kept for API compatibility
    def perform_full_security_check(
        self,
        username: str,
        action: str,
        repository: str,
        entity_type: str,
        entity_id: str,
    ) -> Tuple[bool, str]:
        """Perform comprehensive security check.

        Args:
            username: GitHub username
            action: Action being performed
            repository: Repository name
            entity_type: "issue" or "pr"
            entity_id: Issue/PR number

        Returns:
            Tuple of (allowed, rejection_reason) - reason is empty string if allowed
        """
        if not self.config["enabled"]:
            return True, ""

        # Check user authorization
        if not self._is_user_authorized(username):
            return False, f"User '{username}' is not authorized"

        # Check action authorization
        if not self.is_action_allowed(action):
            return False, f"Action '{action}' is not an allowed action"

        # Check repository
        if not self.is_repository_allowed(repository):
            return False, f"Repository '{repository}' is not authorized"

        # Check rate limit
        if not self.check_rate_limit(username, action):
            return False, "Rate limit exceeded"

        return True, ""

    @property
    def reject_message(self) -> str:
        """Get rejection message."""
        return str(self.config["reject_message"])

File: tools/rust/github-agents-cli/src/security/manager.rs
//! Security manager for GitHub AI Agents.
//!
//! Manages authorization, rate limiting, and security checks for agent operations.

use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::env;
use std::path::Path;
use std::sync::Mutex;
use tracing::{debug, info, warn};

use super::trigger::{is_agent_generated, is_bot_login, is_trigger_response, parse_trigger};
use crate::error::Error;

/// Built-in fallback admin, used when neither a config file nor the
/// `AI_AGENT_DEFAULT_ADMIN` environment variable supplies one.
const BUILTIN_DEFAULT_ADMIN: &str = "AndrewAltimit";

/// Environment variable overriding the built-in default admin(s).
/// Accepts a comma-separated list of usernames.
const DEFAULT_ADMIN_ENV: &str = "AI_AGENT_DEFAULT_ADMIN";

/// Environment variable adding extra allowed users (comma-separated).
const ALLOWED_USERS_ENV: &str = "AI_AGENT_ALLOWED_USERS";

/// Security configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SecurityConfig {
    /// Whether security checks are enabled
    #[serde(default = "default_enabled")]
    pub enabled: bool,

    /// Users with admin privileges
    #[serde(default = "default_agent_admins")]
    pub agent_admins: Vec<String>,

    /// Rate limit window in minutes
    #[serde(default = "default_rate_limit_window")]
    pub rate_limit_window_minutes: u64,

    /// Maximum requests per window
    #[serde(default = "default_rate_limit_max")]
    pub rate_limit_max_requests: usize,

    /// Allowed repositories (empty = all allowed)
    #[serde(default)]
    pub allowed_repositories: Vec<String>,

    /// Message shown when request is rejected
    #[serde(default = "default_reject_message")]
    pub reject_message: String,

    /// Allowed actions
    #[serde(default = "default_allowed_actions")]
    pub allowed_actions: Vec<String>,
}

fn default_enabled() -> bool {
    true
}

/// Resolve the default admin list from an optional env value.
///
/// Split out as a pure function so it can be tested without mutating global
/// process environment (which would race with parallel tests).
fn parse_default_admins(env_value: Option<&str>) -> Vec<String> {
    let admins: Vec<String> = match env_value {
        Some(val) => val
            .split(',')
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect(),
        None => Vec::new(),
    };

    // Fall back to the built-in admin if the env var is unset, empty, or
    // contains only separators/whitespace (e.g. ",,," or "   "). An empty
    // admin list must never be returned: depending on downstream allow-list
    // checks it could lock out every user or allow everyone.
    if admins.is_empty() {
        vec![BUILTIN_DEFAULT_ADMIN.to_string()]
    } else {
        admins
    }
}

fn default_agent_admins() -> Vec<String> {
    parse_default_admins(env::var(DEFAULT_ADMIN_ENV).ok().as_deref())
}

fn default_rate_limit_window() -> u64 {
    60
}

fn default_rate_limit_max() -> usize {
    10
}

fn default_reject_message() -> String {
    "This AI agent only processes requests from authorized users.".to_string()
}

fn default_allowed_actions() -> Vec<String> {
    vec![
        "issue_approved".to_string(),
        "issue_close".to_string(),
        "pr_approved".to_string(),
        "issue_review".to_string(),
        "pr_review".to_string(),
        "issue_summarize".to_string(),
        "pr_summarize".to_string(),
        "issue_debug".to_string(),
        "pr_debug".to_string(),
        "pr_close".to_string(),
    ]
}

impl Default for SecurityConfig {
    fn default() -> Self {
        Self {
            enabled: default_enabled(),
            agent_admins: default_agent_admins(),
            rate_limit_window_minutes: default_rate_limit_window(),
            rate_limit_max_requests: default_rate_limit_max(),
            allowed_repositories: Vec::new(),
            reject_message: default_reject_message(),
            allowed_actions: default_allowed_actions(),
        }
    }
}

/// Where a trigger was found.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TriggerSource {
    /// The issue/PR body
    Body,
    /// The comment at this index (chronological order)
    Comment(usize),
}

/// Trigger information parsed from a comment.
#[derive(Debug, Clone)]
pub struct TriggerInfo {
    /// The action (e.g., "approved", "review")
    pub action: String,
    /// The agent name (optional, may be resolved from board)
    pub agent: Option<String>,
    /// The user who triggered the action
    pub username: String,
    /// Where the trigger was found
    pub source: TriggerSource,
    /// When the triggering comment/body was created (if known)
    pub created_at: Option<DateTime<Utc>>,
}

/// A read-only view of an issue/PR comment used for trigger detection.
///
/// `comments` slices passed to the manager must be in chronological order
/// (the order returned by `gh issue/pr view --json comments`).
#[derive(Debug, Clone, Default)]
pub struct CommentView {
    pub body: String,
    pub author: String,
    pub created_at: Option<DateTime<Utc>>,
}

/// Security manager for AI agent operations.
///
/// Uses interior mutability for the rate limiter so security checks can be
/// performed from `&self` contexts (e.g., async monitor methods).
pub struct SecurityManager {
    config: SecurityConfig,
    allowed_users: HashSet<String>,
    rate_limit_tracker: Mutex<HashMap<String, Vec<DateTime<Utc>>>>,
}

impl SecurityManager {
    /// Create a new security manager with default configuration.
    pub fn new() -> Self {
        let config = SecurityConfig::default();
        let allowed_users = Self::init_allowed_users(&config);

        Self {
            config,
            allowed_users,
            rate_limit_tracker: Mutex::new(HashMap::new()),
        }
    }

    /// Create a security manager from a configuration file.
    ///
    /// Supports two formats:
    /// - Flat `SecurityConfig` (standalone security config file)
    /// - Nested `.agents.yaml` format where security lives under `security:` key
    ///
    /// The file is read through [`super::trust::read_trusted_file`]: in a
    /// PR-triggered run where the PR modifies it, the base-branch version is
    /// used, so a PR cannot add itself to the allow-list.
    pub fn from_config_path(path: &Path) -> Result<Self, Error> {
        let content = super::trust::read_trusted_file(path)?;
        Self::from_yaml_str(&content)
    }

    /// Create a security manager from YAML text (flat or nested format).
    pub fn from_yaml_str(content: &str) -> Result<Self, Error> {
        // A top-level `security` key means the nested `.agents.yaml` format;
        // errors inside it propagate (no silent fallback to defaults).
        let value: serde_yaml::Value = serde_yaml::from_str(content)?;
        let config = match value.get("security") {
            Some(section) if !section.is_null() => {
                info!("Loaded security config from agents YAML (nested format)");
                serde_yaml::from_value::<SecurityConfig>(section.clone())?
            },
            // `security:` present but empty
            Some(_) => SecurityConfig::default(),
            None if value.is_null() => SecurityConfig::default(),
            None => serde_yaml::from_value::<SecurityConfig>(value)?,
        };

        let allowed_users = Self::init_allowed_users(&config);

        Ok(Self {
            config,
            allowed_users,
            rate_limit_tracker: Mutex::new(HashMap::new()),
        })
    }

    /// Create a security manager from a configuration struct.
    #[cfg(test)]
    pub fn from_config(config: SecurityConfig) -> Self {
        let allowed_users = Self::init_allowed_users(&config);
        Self {
            config,
            allowed_users,
            rate_limit_tracker: Mutex::new(HashMap::new()),
        }
    }

    /// Initialize allowed users from config and environment.
    fn init_allowed_users(config: &SecurityConfig) -> HashSet<String> {
        // Case-insensitive comparison: store lowercase
        let mut users: HashSet<String> = config
            .agent_admins
            .iter()
            .map(|u| u.to_lowercase())
            .collect();

        // Add users from environment variable
        if let Ok(env_users) = env::var(ALLOWED_USERS_ENV) {
            for user in env_users.split(',') {
                let user = user.trim();
                if !user.is_empty() {
                    users.insert(user.to_lowercase());
                }
            }
        }

        // Add repository owner (a personal account; for organizations this is
        // the org login, which can never author a comment)
        if let Ok(github_repo) = env::var("GITHUB_REPOSITORY")
            && let Some(owner) = github_repo.split('/').next().filter(|o| !o.is_empty())
        {
            users.insert(owner.to_lowercase());
        }

        // Never allow-list bot accounts or the anonymous/empty login
        users.retain(|u| !u.is_empty() && !is_bot_login(u));

        if !config.enabled {
            warn!("Security checks are DISABLED by configuration (security.enabled: false)");
        }

        debug!("Initialized allowed users: {:?}", users);
        users
    }

    /// Get the rejection message.
    pub fn reject_message(&self) -> &str {
        &self.config.reject_message
    }

    /// Check if a user is authorized.
    ///
    /// Bot accounts and empty logins are never authorized, even when security
    /// checks are otherwise disabled.
    pub fn is_user_allowed(&self, username: &str) -> bool {
        if username.trim().is_empty() || is_bot_login(username) {
            return false;
        }
        if !self.config.enabled {
            return true;
        }
        self.allowed_users.contains(&username.to_lowercase())
    }

    /// Whether `login` may author agent responses that mark a trigger as
    /// handled: allow-listed users, bot accounts, or the account the agent
    /// itself is authenticated as.
    pub fn is_trusted_responder(&self, login: &str, agent_login: Option<&str>) -> bool {
        is_bot_login(login)
            || self.allowed_users.contains(&login.to_lowercase())
            || agent_login.is_some_and(|a| a.eq_ignore_ascii_case(login))
    }

    /// Check if an action is allowed.
    pub fn is_action_allowed(&self, action: &str) -> bool {
        if !self.config.enabled {
            return true;
        }
        self.config.allowed_actions.contains(&action.to_string())
    }

    /// Check if a repository is allowed.
    pub fn is_repository_allowed(&self, repository: &str) -> bool {
        if !self.config.enabled {
            return true;
        }
        // Empty list means all repositories are allowed
        if self.config.allowed_repositories.is_empty() {
            return true;
        }
        self.config
            .allowed_repositories
            .iter()
            .any(|r| r.eq_ignore_ascii_case(repository))
    }

    /// Check and update rate limit for a user/action combination.
    pub fn check_rate_limit(&self, username: &str, action: &str) -> bool {
        if !self.config.enabled {
            return true;
        }

        let key = format!("{}:{}", username.to_lowercase(), action);
        let now = Utc::now();
        let window = Duration::minutes(self.config.rate_limit_window_minutes as i64);
        let cutoff = now - window;

        let mut tracker = self
            .rate_limit_tracker
            .lock()
            .unwrap_or_else(|e| e.into_inner());

        // Clean old entries
        let requests = tracker.entry(key.clone()).or_default();
        requests.retain(|t| *t > cutoff);

        // Check limit
        if requests.len() >= self.config.rate_limit_max_requests {
            warn!(
                "Rate limit exceeded for {}:{} ({} requests in {} minutes)",
                username,
                action,
                requests.len(),
                self.config.rate_limit_window_minutes
            );
            return false;
        }

        // Record request
        requests.push(now);
        true
    }

    /// Check for a valid trigger in issue/PR data.
    ///
    /// Iterates comments in reverse order so the **latest** matching trigger
    /// from an authorized user wins, allowing newer directives to supersede
    /// older ones. Falls back to the issue/PR body only if no comment
    /// triggers exist. Comments and bodies generated by this tool are never
    /// considered, which prevents prompt-injected agent output from
    /// self-authorizing when the agent posts under an allow-listed account.
    pub fn check_trigger_comment(
        &self,
        body: &str,
        author: &str,
        body_created_at: Option<DateTime<Utc>>,
        comments: &[CommentView],
    ) -> Option<TriggerInfo> {
        for (index, comment) in comments.iter().enumerate().rev() {
            if is_agent_generated(&comment.body) || !self.is_user_allowed(&comment.author) {
                continue;
            }
            if let Some(trigger) = parse_trigger(&comment.body) {
                return Some(TriggerInfo {
                    action: trigger.action,
                    agent: trigger.agent,
                    username: comment.author.clone(),
                    source: TriggerSource::Comment(index),
                    created_at: comment.created_at,
                });
            }
        }

        if !is_agent_generated(body)
            && self.is_user_allowed(author)
            && let Some(trigger) = parse_trigger(body)
        {
            return Some(TriggerInfo {
                action: trigger.action,
                agent: trigger.agent,
                username: author.to_string(),
                source: TriggerSource::Body,
                created_at: body_created_at,
            });
        }

        None
    }

    /// Whether the agent already responded to `trigger`.
    ///
    /// A trigger is handled once a monitor reply (see
    /// [`super::trigger::TRIGGER_RESPONSE_MARKER`]) from a trusted responder
    /// appears after it. This makes monitor runs idempotent: the same
    /// `[Approved]` comment is never acted on twice, and security rejections
    /// are not re-posted on every polling cycle.
    pub fn is_trigger_handled(
        &self,
        trigger: &TriggerInfo,
        comments: &[CommentView],
        agent_login: Option<&str>,
    ) -> bool {
        let start = match trigger.source {
            TriggerSource::Body => 0,
            TriggerSource::Comment(index) => index + 1,
        };
        comments.iter().skip(start).any(|c| {
            is_trigger_response(&c.body) && self.is_trusted_responder(&c.author, agent_login)
        })
    }

    /// Perform a comprehensive security check.
    ///
    /// Returns (allowed, rejection_reason).
    pub fn perform_full_security_check(
        &self,
        username: &str,
        action: &str,
        repository: &str,
    ) -> (bool, String) {
        if !self.config.enabled {
            return (true, String::new());
        }

        // Check user authorization
        if !self.is_user_allowed(username) {
            return (false, format!("User '{}' is not authorized", username));
        }

        // Check action authorization
        if !self.is_action_allowed(action) {
            return (
                false,
                format!("Action '{}' is not an allowed action", action),
            );
        }

        // Check repository
        if !self.is_repository_allowed(repository) {
            return (
                false,
                format!("Repository '{}' is not authorized", repository),
            );
        }

        // Check rate limit
        if !self.check_rate_limit(username, action) {
            return (false, "Rate limit exceeded".to_string());
        }

        (true, String::new())
    }

    /// Get list of allowed users (for debugging/info).
    pub fn allowed_users(&self) -> Vec<String> {
        self.allowed_users.iter().cloned().collect()
    }
}

impl Default for SecurityManager {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_user_allowed_case_insensitive() {
        let manager = SecurityManager::new();
        assert!(manager.is_user_allowed("AndrewAltimit"));
        assert!(manager.is_user_allowed("andrewaltimit"));
        assert!(manager.is_user_allowed("ANDREWALTIMIT"));
    }

    #[test]
    fn test_rate_limit() {
        let manager = SecurityManager::from_config(SecurityConfig {
            rate_limit_max_requests: 2,
            ..Default::default()
        });

        assert!(manager.check_rate_limit("user", "action"));
        assert!(manager.check_rate_limit("user", "action"));
        assert!(!manager.check_rate_limit("user", "action"));
    }

    #[test]
    fn test_full_security_check() {
        let manager = SecurityManager::new();
        let (allowed, reason) = manager.perform_full_security_check(
            "AndrewAltimit",
            "issue_approved",
            "AndrewAltimit/repo",
        );
        assert!(allowed);
        assert!(reason.is_empty());
    }

    #[test]
    fn test_parse_default_admins() {
        // Unset / empty -> built-in fallback
        assert_eq!(parse_default_admins(None), vec![BUILTIN_DEFAULT_ADMIN]);
        assert_eq!(parse_default_admins(Some("")), vec![BUILTIN_DEFAULT_ADMIN]);
        assert_eq!(
            parse_default_admins(Some("   ")),
            vec![BUILTIN_DEFAULT_ADMIN]
        );

        // Only separators / whitespace -> built-in fallback (never an empty list)
        assert_eq!(parse_default_admins(Some(",")), vec![BUILTIN_DEFAULT_ADMIN]);
        assert_eq!(
            parse_default_admins(Some(",,,")),
            vec![BUILTIN_DEFAULT_ADMIN]
        );
        assert_eq!(
            parse_default_admins(Some(" , , ")),
            vec![BUILTIN_DEFAULT_ADMIN]
        );

        // Single override
        assert_eq!(parse_default_admins(Some("Alice")), vec!["Alice"]);

        // Comma-separated list with surrounding whitespace and empties
        assert_eq!(
            parse_default_admins(Some(" Alice , Bob ,, ")),
            vec!["Alice", "Bob"]
        );
    }

    #[test]
    fn test_unauthorized_user() {
        let manager = SecurityManager::new();
        let (allowed, reason) =
            manager.perform_full_security_check("unknown_user", "issue_approved", "owner/repo");
        assert!(!allowed);
        assert!(reason.contains("not authorized"));
    }

    #[test]
    fn test_agents_yaml_nested_parsing() {
        let yaml = r#"
enabled_agents:
  - claude
security:
  agent_admins:
    - TestAdmin
  rate_limit_window_minutes: 30
"#;
        let m = SecurityManager::from_yaml_str(yaml).unwrap();
        assert_eq!(m.config.agent_admins, vec!["TestAdmin"]);
        assert_eq!(m.config.rate_limit_window_minutes, 30);
        assert!(m.is_user_allowed("testadmin"));
    }

    #[test]
    fn test_agents_yaml_without_security_key_uses_flat_defaults() {
        let yaml = r#"
enabled_agents:
  - claude
"#;
        let m = SecurityManager::from_yaml_str(yaml).unwrap();
        assert_eq!(m.config.rate_limit_window_minutes, 60);
    }

    #[test]
    fn test_flat_config_format() {
        let yaml = r#"
agent_admins:
  - FlatAdmin
rate_limit_window_minutes: 15
"#;
        let m = SecurityManager::from_yaml_str(yaml).unwrap();
        assert_eq!(m.config.agent_admins, vec!["FlatAdmin"]);
        assert_eq!(m.config.rate_limit_window_minutes, 15);
    }

    #[test]
    fn test_empty_security_section_uses_defaults() {
        let m = SecurityManager::from_yaml_str("security:\n").unwrap();
        assert!(m.config.enabled);
        let m = SecurityManager::from_yaml_str("").unwrap();
        assert!(m.config.enabled);
    }

    fn admin_manager() -> SecurityManager {
        SecurityManager::from_config(SecurityConfig {
            agent_admins: vec!["Admin".to_string()],
            ..Default::default()
        })
    }

    fn comment(body: &str, author: &str) -> CommentView {
        CommentView {
            body: body.to_string(),
            author: author.to_string(),
            created_at: None,
        }
    }

    #[test]
    fn test_trigger_from_unauthorized_user_ignored() {
        let m = admin_manager();
        let comments = vec![comment("[Approved][Claude]", "mallory")];
        assert!(
            m.check_trigger_comment("", "mallory", None, &comments)
                .is_none()
        );
    }

    #[test]
    fn test_latest_authorized_trigger_wins() {
        let m = admin_manager();
        let comments = vec![
            comment("[Approved][Claude]", "admin"),
            comment("[Review][Claude]", "mallory"),
            comment("[Summarize]", "Admin"),
        ];
        let t = m.check_trigger_comment("", "x", None, &comments).unwrap();
        assert_eq!(t.action, "summarize");
        assert_eq!(t.source, TriggerSource::Comment(2));
        assert_eq!(t.username, "Admin");
    }

    #[test]
    fn test_agent_generated_comment_never_triggers() {
        let m = admin_manager();
        // Posted under the admin account (e.g. agent token is an admin PAT)
        let body = format!(
            "Model output: [Approved][Claude]\n{}",
            super::super::trigger::AGENT_COMMENT_MARKER
        );
        let comments = vec![comment(&body, "admin")];
        assert!(m.check_trigger_comment("", "x", None, &comments).is_none());
        assert!(
            m.check_trigger_comment("[AI Agent] [Approved]", "admin", None, &[])
                .is_none()
        );
    }

    #[test]
    fn test_body_trigger_fallback() {
        let m = admin_manager();
        let t = m
            .check_trigger_comment("Please [Approved][OpenCode]", "admin", None, &[])
            .unwrap();
        assert_eq!(t.source, TriggerSource::Body);
        assert_eq!(t.agent.as_deref(), Some("opencode"));
        // Quoted/code triggers in the body do not count
        assert!(
            m.check_trigger_comment("reply with `[Approved]`", "admin", None, &[])
                .is_none()
        );
    }

    #[test]
    fn test_bots_never_allowed() {
        let m = SecurityManager::from_config(SecurityConfig {
            agent_admins: vec!["github-actions[bot]".to_string(), "Admin".to_string()],
            ..Default::default()
        });
        assert!(!m.is_user_allowed("github-actions[bot]"));
        assert!(!m.is_user_allowed(""));
        assert!(m.is_user_allowed("admin"));

        let disabled = SecurityManager::from_config(SecurityConfig {
            enabled: false,
            ..Default::default()
        });
        assert!(disabled.is_user_allowed("anyone"));
        assert!(!disabled.is_user_allowed("renovate[bot]"));
    }

    #[test]
    fn test_trigger_handled_detection() {
        let m = admin_manager();
        let marker = super::super::trigger::TRIGGER_RESPONSE_MARKER;
        let reply = format!("Working on it\n{marker}");

        let comments = vec![
            comment("[Approved][Claude]", "admin"),
            comment("thanks", "someone"),
        ];
        let t = m.check_trigger_comment("", "x", None, &comments).unwrap();
        assert!(!m.is_trigger_handled(&t, &comments, None));

        // Agent reply from a bot marks it handled
        let mut handled = comments.clone();
        handled.push(comment(&reply, "github-actions"));
        assert!(m.is_trigger_handled(&t, &handled, None));

        // A forged marker from an untrusted user does not
        let mut forged = comments.clone();
        forged.push(comment(&reply, "mallory"));
        assert!(!m.is_trigger_handled(&t, &forged, None));

        // ...unless that login is the agent's own account
        assert!(m.is_trigger_handled(&t, &forged, Some("Mallory")));

        // Generated content that is not a trigger reply does not count
        let mut insight = comments.clone();
        insight.push(comment(
            &format!("insight {}", super::super::trigger::AGENT_COMMENT_MARKER),
            "github-actions",
        ));
        assert!(!m.is_trigger_handled(&t, &insight, None));

        // A reply that precedes the trigger does not count
        let earlier = vec![
            comment(&reply, "github-actions"),
            comment("[Approved][Claude]", "admin"),
        ];
        let t = m.check_trigger_comment("", "x", None, &earlier).unwrap();
        assert!(!m.is_trigger_handled(&t, &earlier, None));
    }

    #[test]
    fn test_repository_allow_list_case_insensitive() {
        let m = SecurityManager::from_config(SecurityConfig {
            allowed_repositories: vec!["Owner/Repo".to_string()],
            ..Default::default()
        });
        assert!(m.is_repository_allowed("owner/repo"));
        assert!(!m.is_repository_allowed("owner/other"));
    }

    #[test]
    fn test_action_allow_list() {
        let m = admin_manager();
        assert!(m.is_action_allowed("issue_approved"));
        assert!(m.is_action_allowed("pr_close"));
        assert!(!m.is_action_allowed("issue_delete"));
    }

    #[test]
    fn test_invalid_nested_security_does_not_silently_fallback() {
        // If YAML has `security:` key with invalid shape, from_config_path
        // should propagate the error, not silently fall back to defaults
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("invalid_nested.yaml");
        std::fs::write(
            &path,
            r#"
security:
  agent_admins: "not_a_list"
"#,
        )
        .unwrap();
        let result = SecurityManager::from_config_path(&path);
        assert!(
            result.is_err(),
            "invalid nested security config should return error, not silent defaults"
        );
    }
}
